File: ios/Packages/SimplicityServices/Sources/SimplicityServices/ReflectionService.swift
import Foundation
import Mockable
import SimplicityApi

/// The clinician's private journal.
///
/// Note the absence of an organisation anywhere in this protocol. A reflection belongs to a
/// person, not to a workplace — the endpoints are under `/api/v1/me` — and this must not acquire
/// an `orgId` for symmetry with the other services.
@Mockable
public protocol ReflectionService: AnyObject, Sendable {
    func list(query: String?) async throws -> [ReflectionResponse]
    func write(title: String?, body: String) async throws -> ReflectionResponse
    func edit(id: UUID, title: String?, body: String) async throws -> ReflectionResponse
    func delete(id: UUID) async throws
}

File: android/services/src/main/kotlin/io/simplicity/training/services/ReflectionService.kt
package io.simplicity.training.services

import io.simplicity.training.api.apis.ReflectionsApi
import io.simplicity.training.api.models.ReflectionResponse
import io.simplicity.training.api.models.WriteReflectionRequest
import java.util.UUID

/**
 * A clinician's private journal.
 *
 * Nothing here is organisation-scoped, deliberately: a reflection belongs to its author and to
 * nobody else, which is the promise the product makes and the reason the endpoints take no org id.
 */
interface ReflectionService {
    suspend fun list(query: String? = null): List<ReflectionResponse>
    suspend fun write(title: String?, body: String): ReflectionResponse
    suspend fun edit(id: UUID, title: String?, body: String): ReflectionResponse
    suspend fun delete(id: UUID)
}

class ReflectionServiceImpl(private val api: ReflectionsApi) : ReflectionService {

    override suspend fun list(query: String?) = api.listReflections(query?.takeIf { it.isNotBlank() }).unwrap()

    override suspend fun write(title: String?, body: String) =
        api.writeReflection(WriteReflectionRequest(body = body, title = title)).unwrap()

    override suspend fun edit(id: UUID, title: String?, body: String) =
        api.editReflection(id, WriteReflectionRequest(body = body, title = title)).unwrap()

    override suspend fun delete(id: UUID) {
        val response = api.deleteReflection(id)
        if (!response.isSuccessful) throw ApiFailure(response.code(), "Delete failed")
    }
}
